src/meta/src/barrier/mod.rs

```diff
@@ -162,6 +182,34 @@ pub struct GlobalBarrierManager<S: MetaStore> {
     metrics: Arc<MetaMetrics>,
 
     env: MetaSrvEnv<S>,
+
+    /// Signal whether to start recovery
+    is_recovery: Arc<AtomicBool>,
```

Comment:
`is_recovery` and `is_build_actor` is a kind of state of the `GlobalBarrierManager`. The two states is only read in the `can_inject_barrier` method, and `can_inject_barrier` is only used in the worker loop of barrier manager. 

The state will be modified in an atomic variable, and the value can be modified at any time and it's likely to cause some concurrency issue. So instead of explicitly setting the two state variable, we can maintain a state machine in the barrier manager worker loop. In the first state, for example, named `PauseInjection`, we only `select!` from the recovery future, and in the normal running state, we `select!` from all current futures. We may introduce a new channel to accept the signal to transform the state.

Reply:
+1 for maintining a state machine.